pkg/tsdb/interval.go

```diff
@@ -6,25 +6,52 @@ import (
 )
 
 var (
-	defaultRes  int64         = 1500
-	minInterval time.Duration = 1 * time.Millisecond
-	year        time.Duration = time.Hour * 24 * 365
-	day         time.Duration = time.Hour * 24 * 365
+	defaultRes         int64         = 1500
+	defaultMinInterval time.Duration = 1 * time.Millisecond
+	year               time.Duration = time.Hour * 24 * 365
+	day                time.Duration = time.Hour * 24 * 365
 )
 
 type Interval struct {
 	Text  string
 	Value time.Duration
 }
 
-func CalculateInterval(timerange *TimeRange) Interval {
-	interval := time.Duration((timerange.MustGetTo().UnixNano() - timerange.MustGetFrom().UnixNano()) / defaultRes)
+type intervalCalculator struct {
+	minInterval time.Duration
+}
+
+type IntervalOptions struct {
+	MinInterval time.Duration
+}
+
+func NewIntervalCalculator(opt *IntervalOptions) *intervalCalculator {
+	if opt == nil {
+		opt = &IntervalOptions{}
+	}
+
+	calc := &intervalCalculator{}
+
+	if opt.MinInterval == 0 {
+		calc.minInterval = defaultMinInterval
+	} else {
+		calc.minInterval = opt.MinInterval
+	}
+
+	return calc
+}
+
+func (ic *intervalCalculator) Calculate(timerange *TimeRange) Interval {
+	to := timerange.MustGetTo().UnixNano()
+	from := timerange.MustGetFrom().UnixNano()
+	interval := time.Duration((to - from) / defaultRes)
 
-	if interval < minInterval {
-		return Interval{Text: formatDuration(minInterval), Value: interval}
+	if interval < ic.minInterval {
+		return Interval{Text: formatDuration(ic.minInterval), Value: ic.minInterval}
 	}
 
-	return Interval{Text: formatDuration(roundInterval(interval)), Value: interval}
```

Comment:
Not sure why we interval directly before. Can't think of any reason where text and value shouldn't align?

Reply:
what do you mean interval directly? the Text version is for example "10s" and the value I think is the raw duration value (in go duration type). The text can be used in query languages that support it I think, the the raw value can be used otherwise. 

For example InfluxDB uses the Text version: 
https://github.com/grafana/grafana/blob/master/pkg/tsdb/influxdb/query.go#L38

Reply:
We only used the rounded value for the `Text` property. I think both `Text` and `Value` should be rounded.